### ISSUE TITLE
Improve select and select! performance with `Not`

### ISSUE THREAD

**Issue by bkamins** [JuliaData/DataFrames.jl#1861]
`select` and `select!` internally removes columns. Therefore it is relatively easy to add special methods for `Not` column selector that will have a much better performance than current implementation that does `setdiff` twice.

**Comment by Ellipse0934** [comment]
Hi,
I tried really hard to improve performance using the `Not` selector but could not do it. I could achieve a little worse than the current performance. Unfortunately `Not` is having too many allocations internally which gets worse with the DataFrame's size. (There is an issue at InvInd/issues which mentions this)
I did manage to improve the performance significantly by rewriting `select!`. Currently we generate a list of columns to be removed then remove them sequentially and each removal is  O(df.columns). After which there is the cost of using `permute!`.
Instead how about swapping in-place in df.columns as each swap is O(1) and we need to make |inds| swaps at max. Then just resizing df.columns to |inds| to purge unwanted cols. (And same swapping for colindices)
```julia
function mselect!(df::DataFrame, inds::AbstractVector{Int})
    if isempty(inds)
        empty!(_columns(df))
        empty!(index(df))
        return df
    end
    indmin, indmax = extrema(inds)
    if indmin < 1
        throw(ArgumentError("indices must be positive"))
    end
    if indmax > ncol(df)
        throw(ArgumentError("indices must not be greater than number of columns"))
    end
    if !allunique(inds)
        throw(ArgumentError("indices must not contain duplicates"))
    end
    
    itr = 1
    ord = collect(1:length(_columns(df)))
    pos = copy(ord)
    for i = inds
            ord[itr], ord[pos[i]] = ord[pos[i]], ord[itr]
            _columns(df)[itr], _columns(df)[pos[i]] = _columns(df)[pos[i]], _columns(df)[itr]
            index(df).names[itr], index(df).names[pos[i]] = index(df).names[pos[i]], index(df).names[itr]
            pos[ord[pos[i]]], pos[ord[itr]] = pos[ord[itr]], pos[ord[pos[i]]]
            itr += 1
    end
    l = length(index(df).names)
    for i = 1:length(inds)
          index(df).lookup[index(df).names[i]] = i
    end
    for i = (length(inds) + 1):l
        delete!(index(df).lookup,index(df).names[i])
    end
   resize!(_columns(df), length(inds))
   resize!(index(df).names, length(inds)) 
  df
end
```
Performance Improvement:

```julia
cols = 90000
df = DataFrame(rand(500, cols))
dfc = copy(df)
indx = randperm(cols)[1:Int.(cols/2)];
@time select!(df, indx);
    > 19.674033 seconds (107 allocations: 11.190 MiB)
@time mselect!(dfc, indx)
    >  0.013488 seconds (40 allocations: 4.376 MiB)
```
Should I send a PR ?

**Comment by bkamins** [comment]
An excellent work on it - thank you. Indeed the `Not` inefficiency is dwarfed by other operations so we can probably neglect it in practice as you noted.

Regarding your change. Yes - it would be very good to add a PR for this. I have done some benchmarking and I think that the change along this line:
```
function select2!(df::DataFrame, inds::AbstractVector{Int})
    if isempty(inds)
        empty!(DataFrames._columns(df))
        empty!(DataFrames.index(df))
        return df
    end
    indmin, indmax = extrema(inds)
    if indmin < 1
        throw(ArgumentError("indices must be positive"))
    end
    if indmax > ncol(df)
        throw(ArgumentError("indices must not be greater than number of columns"))
    end
    if !allunique(inds)
        throw(ArgumentError("indices must not contain duplicates"))
    end
    targetnames = _names(df)[inds]
    targetcolumns = _columns(df)[inds]
    empty!(_columns(df))
    append!(_columns(df), targetcolumns)
    x = index(df)
    empty!(x)
    append!(_names(x), targetnames)
    for (i, n) in enumerate(x.names)
        x.lookup[n] = i
    end
    df
end
```
should achieve the same effect as what you propose, but it seems to be a bit faster (and the code is simpler, so it would be easier to maintain in the future). I have not done extensive tests of it, so there might be some bug there, so it would require some testing.

In summary - yes, please open a PR for this, but consider my proposal in it (or maybe you will find yet another approach that is fast and has a relatively simple code). Thank you!

**Comment by Ellipse0934** [comment]
I tried but unfortunately could not improve performance and the PR I have submitted is your code copy/pasted. I could not find any bugs upon manually testing. 
There might be opportunities to improve but it's difficult and the improvement probably isn't worth it for now.

**Comment by bkamins** [comment]
Fixed by #1985.
